mh/state.py

```diff
@@ -589,9 +792,518 @@ def match_city(city, fields):
+        self.wfile.write(self.pack_data(
+            data, packet_id
+        ))
+
+    def pack_data(self, _data, packet_id):
+        data = struct.pack(">II", len(_data), packet_id)
+        data += _data
+        return data
+
+    def is_finished(self):
+        return self.finished
+
+    def on_exception(self, e):
+        self.finish()
+
+    def find_handler(self, packet):
+        self.handler_functions[packet[1]](packet[0], packet[2])
+
+    def RecvFriendlyHello(self, server_id, data):
+        self.cache.debug("Recieved a friendly hello from {}!".format(
+            server_id
+        ))
+        self.cache.register_handler(server_id, self)
+        self.ReqConnectionInfo()
+
+    def ReqConnectionInfo(self):
+        self.cache.debug("Requesting connection info.")
+        self.send_packet(PacketTypes.ReqConnectionInfo, b"")
+
+    def RecvConnectionInfo(self, server_id, data):
+        self.cache.debug("Recieved connection info sized {} from {}".format(
+            len(data), server_id
+        ))
+        server = Server.deserialize(json.loads(data))
+        self.cache.servers[server_id] = server
+        self.cache.update_players()
+
+    def RecvReqConnectionInfo(self, server_id, data):
+        requested_server_id, = struct.unpack(">H", data)
+        self.cache.debug("Recieved request for data of Server {}.".format(
+            requested_server_id
+        ))
+        if server_id in self.cache.servers:
+            data = json.dumps(self.cache.servers[server_id].serialize())
+            self.SendConnectionInfo(data)
+
+    def SendConnectionInfo(self, data):
+        self.send_packet(PacketTypes.SendConnectionInfo, data)
+
+    def RecvReqServerRefresh(self, server_id, data):
+        self.cache.debug("Recieved server refresh request from Server {}.".format(
+            server_id
+        ))
+        for _server_id in self.cache.servers:
+            data = struct.pack(">H", _server_id)
+            data += json.dumps(self.cache.servers[_server_id].serialize())
+            self.SendConnectionInfo(data)
+
+    def RecvSessionInfo(self, server_id, data):
+        dest_server_id, = struct.unpack(">H", data[:2])
+        self.cache.debug("Recieved session data from Server {} \
+                            bound for Server {}.".format(
+            server_id, dest_server_id
+        ))
+        from mh.session import Session
+        self.cache.update_player_record(Session.deserialize(json.loads(data[2:])))
+        self.cache.get_handler(dest_server_id).SendSessionInfo(data[2:])
+
+    def SendSessionInfo(self, ser_session):
+        self.send_packet(PacketTypes.SessionInfo, ser_session)
+
+    def finish(self):
+        if self.finished:
+            return
+
+        self.finished = True
+
+        try:
+            self.wfile.close()
+        except Exception:
+            pass
+
+        try:
+            self.rfile.close()
+        except Exception:
+            pass
+
+
+class RemoteConnectionHandler(object):
```

Comment:
Ditto, it has to be in its own file. Do you have a solid reason not to inherit from an existing python class?

Reply:
Moved to `other/cache.py`